[docs] blue-green deploys with pinned schemas (#2850)
apollographql/federation: docs/source/managed-federation/setup.mdx

This option is specific to _non_-managed federation, in which supergraph schema composition is performed via the Rover CLI.

> With managed federation, composition can instead be performed by _GraphOS_, where the gateway regularly polls for an updated schema. This enables you to add, remove, and modify your subgraphs _without_ needing to restart your gateway.

Reviewer: Non-blocking, just curious about the change from "is" to "can" across this page. With managed federation isn't the point that composition _is_ instead handled by GraphOS?

Author: The distinction is that "managed federation" in general manages one or all aspects of CI/CD for federation, and composition may or may not be managed (`can` vs. `is).

That said, you're right that this sentence is focused on what GraphOS provides. Rewriting to `With federation managed by GraphOS, composition is performed and ...`

> Remove the `supergraphSdl` argument from your `ApolloGateway` constructor entirely:

Reviewer: ```suggestion
To use managed federation, ensure your `ApolloGateway` constructor doesn't include the `supergraphSdl` argument:
```

```js
const gateway = new ApolloGateway();
```

When running your gateway in an environment where outbound traffic to the internet is restricted, consult the [directions for configuring a proxy](/apollo-server/security/proxy-configuration) within Apollo Server.

</ExpansionPanel>

If you've already set up Apollo Federation _without_ connecting your router to GraphOS, you're probably passing the [`--supergraph` (or `-s`) option](/router/configuration/overview/#-s----supergraph) to the Apollo Router's startup command:

```sh
./router --config ./router.yaml --supergraph ./your-local-supergraph.graphql
```

> The `--supergraph` option applies in scenarios where the supergraph schema for the router isn't retrieved by polling Apollo Uplink:

Reviewer: ```suggestion
The `--supergraph` option applies in scenarios where the router doesn't retrieve the supergraph schema by polling Apollo Uplink. These include:
```

> * For enabling different production environments (such as blue-green deployments) to refer to a single graph variant, the supergraph schema for each production environment can be composed and retrieved by a workflow using Platform APIs, so that the routers of each environment use the supergraph schema corresponding to the environment. 

Reviewer: I found this a little hard to follow and tried to break it up to make it more readable. That said, it seems a little contradictory to me. On the one hand, we seem to be saying that different production environments point to the same variant (and therefore the same schema?). On the other hand, we're saying each environments router can point to its (own?) corresponding supergraph schema? 

```suggestion
* **Enabling different production environments to refer to a single graph variant**, for example, for [blue-green](https://en.wikipedia.org/wiki/Blue%E2%80%93green_deployment): Workflows using the [Platform API](/graphos/platform-api/) compose and retrieve the supergraph schema for each production environment. This enables each environment's router to use the environment's supergraph schema. 
```

> * For fully unmanaged federation, the Rover CLI composes the supergraph schema, and the file path of the schema is passed to the router. 

Reviewer: ```suggestion
* **Unmanaged federation**: The Rover CLI composes the supergraph schema and passes the schema's file path to the router via the `--supergraph` option.
```

> * **Enabling different deployment environments to refer to a single graph variant**, for example, for [blue-green deployment](https://en.wikipedia.org/wiki/Blue%E2%80%93green_deployment): Use a workflow with the [Platform API](/graphos/platform-api/) to compose and retrieve the supergraph schema for each environment. This enables each environment's router to use the environment's supergraph schema, and then a single graph variant can capture and analyze the operations and traffic of either environment when deployed.

Reviewer: Whoops, previously suggested the Wikipedia link before seeing the new section where you nicely explain the concept of blue-green deployments.

```suggestion
* **Enabling different deployment environments to refer to a single graph variant**, for example, for [blue-green deployment](./deployment/#example-blue-green-deployment): Use a workflow with the [Platform API](/graphos/platform-api/) to compose and retrieve the supergraph schema for each environment. This enables each environment's router to use the environment's supergraph schema, and then a single graph variant can capture and analyze the operations and traffic of either environment when deployed.
```

> For other scenarios where the supergraph schema is gotten from Apollo Uplink, _remove_ the `--supergraph` option (but leave `--config` if it's present):

Reviewer: ```suggestion
For other scenarios where the router retrieves the supergraph schema is from Apollo Uplink, _remove_ the `--supergraph` option (but leave `--config` if it's present):
```
